python/datafusion/functions.py

```diff
@@ -15,9 +15,1731 @@
+    """
+    Returns `base` raised to the power of `exponent`.
+    This is an alias of `power`.
+    """
+    return power(base, exponent)
+
+
+def radians(arg: Expr) -> Expr:
+    """
+    Converts the argument from degrees to radians.
+    """
+    return Expr(f.radians(arg.expr))
+
+
+def regexp_like(string: Expr, regex: Expr, flags: Expr | None = None) -> Expr:
+    """
+    Tests a string using a regular expression returning true if at
+    least one match, false otherwise.
+    """
+    if flags is not None:
+        flags = flags.expr
+    return Expr(f.regexp_like(string.expr, regex.expr, flags))
+
+
+def regexp_match(string: Expr, regex: Expr, flags: Expr | None = None) -> Expr:
+    """
+    Returns an array with each element containing the leftmost-first
+    match of the corresponding index in `regex` to string in `string`
+
+    If there is no match, the list element is NULL.
+
+    If a match is found, and the pattern contains no capturing parenthesized subexpressions,
+    then the list element is a single-element [`GenericStringArray`] containing the substring
+    matching the whole pattern.
+
+    If a match is found, and the pattern contains capturing parenthesized subexpressions, then the
+    list element is a [`GenericStringArray`] whose n'th element is the substring matching
+    the n'th capturing parenthesized subexpression of the pattern.
+    """
+
+    # TODO VALIDATE THIS IS CORRECT FOR DATAFRAME RESULTS
+    if flags is not None:
+        flags = flags.expr
+    return Expr(f.regexp_match(string.expr, regex.expr, flags))
+
+
+def regexp_replace(
+    string: Expr, pattern: Expr, replacement: Expr, flags: Expr | None = None
+) -> Expr:
+    """
+    Replaces substring(s) matching a PCRE-like regular expression.
+
+    The full list of supported features and syntax can be found at
+    <https://docs.rs/regex/latest/regex/#syntax>
+
+    Supported flags with the addition of 'g' can be found at
+    <https://docs.rs/regex/latest/regex/#grouping-and-flags>
+    """
+    if flags is not None:
+        flags = flags.expr
+    return Expr(f.regexp_replace(string.expr, pattern.expr, replacement.expr, flags))
+
+
+def repeat(string: Expr, n: Expr) -> Expr:
+    """
+    Repeats the `string` to `n` times.
+    """
+    return Expr(f.repeat(string.expr, n.expr))
+
+
+def replace(string: Expr, from_val: Expr, to_val: Expr) -> Expr:
+    """
+    Replaces all occurrences of `from` with `to` in the `string`.
+    """
+    return Expr(f.replace(string.expr, from_val.expr, to_val.expr))
+
+
+def reverse(arg: Expr) -> Expr:
+    """
+    Reverse the string argument.
+    """
+    return Expr(f.reverse(arg.expr))
+
+
+def right(string: Expr, n: Expr) -> Expr:
+    """
+    Returns the last `n` characters in the `string`.
+    """
+    return Expr(f.right(string.expr, n.expr))
+
+
+def round(arg: Expr) -> Expr:
+    """
+    Round the argument to the nearest integer.
+    """
+    return Expr(f.round(arg.expr))
```

Comment:
Looks like this is missing its optional second parameter based on the test failures

```
col_name = 'revenue', col_type = Decimal128Type(decimal128(38, 4))

    def df_selection(col_name, col_type):
        if col_type == pa.float64() or isinstance(col_type, pa.Decimal128Type):
>           return F.round(col(col_name), lit(2)).alias(col_name)
E           TypeError: round() takes 1 positional argument but 2 were given
```

Maybe something like this, similar to rpad?
```suggestion
def round(arg: Expr, decimal_places: Expr | None = None) -> Expr:
    """
    Round the argument to the nearest integer. If decimal_places is specified, rounds to the given number of decimal places.
    """
    decimal_places = decimal_places if decimal_places is not None else Expr.literal(0)
    return Expr(f.round(arg.expr, decimal_places.expr))
```
?

Reply:
In these situations, does it make sense to specify `None` as the default value, or embed `Expr.literal(0)` directly? (Is Expr mutable - could that cause issues similar to the use of `{}` as a default value above? If not, I think this could help better document the default value that's used, in the type hint directly, rather than just implied in the documentation.)

Reply:
Updated to your recommendation.